### PR TITLE
Enable `clippy_pedantic` and other cleanup.

### PR DESCRIPTION
**Pull request by arsing**
- All crates:

  - Clippy lints:

    - Use `Self` in function return types.
    - Identifiers in doc comments are now wrapped in code fences.
    - `Option.map().unwrap_or{_else}()` -> `Option.map_or{_else}()`
    - `Result.map().unwrap_or_else()` -> `match Result { }`
    - `for foo in bar.iter()` -> `for foo in &bar`
    - Combine match arms with common expressions.
    - Break up long numeric literals with `_` (`3_221_225_472`)
    - Separate number from type suffix with `_` (`0_usize`)
    - `foo.len() > 0` -> `!foo.is_empty()`
    - Some tests would `match` a `Result` and `panic!`. Replaced with
      `expect{_err}`
    - Constants of `&'static str` type can be written as just `&str` since
      the `'static` lifetime is implicit.

  - Other cleanup:

    - Functions that returned `&String`/`&Vec`/`Option<&String>`/`Option<&Vec>`
      now return `&str`/`&[]`/`Option<&str>`/`Option<&[]>`
    - Derive `Copy` on types that can be `Copy`. clippy does not suggest this,
      but it *does* lint for `Copy` types that are `clone()`d or taken
      by reference.
    - `Iterator.nth(0)` -> `Iterator.next()`

- `docker`:

  - Rename `Error::ApiError` variant to `Error::Api`

- `edgelet-docker`

  - Avoid potential overflow converting u64 to i32.

- `edgelet-hsm`

  - `Certificate` and `TpmKey` now impl `Debug`, because some tests that
    matched on `Result`s and panicked now use `Result.expect{_err}()` which
    requires the `Debug` bound.

- `edgelet-http-workload`

  - Some `static` `&'static str` changed to `const`

- `hsm-rs`

  - Some negative tests stored the result of functions in bindings, even though
    these functions returned `()`. Removed the bindings. Positive tests which
    asserted that the functions returned `()` have not been changed.
  - `0_isize as *mut c_void` -> `std::ptr::null_mut()`

- `hyper-named-pipe`

  - Clarified the reason for the workaround that `extern crate failure`
    needs to come before `extern crate edgelet_utils`.

- `iotedge`

  - Some byte arrays using hex literals changed to use byte character literals
    and byte string literals.
  - `test_read` did not assert the result of `Read::read`. Changed to
    `read_exact`

- `iothubservice`

  - `AuthMechanism::_type` -> `AuthMechanism::type_` since leading `_` has a
    special meaning in Rust.
  - `Ok(()) as Result<(), Error>` -> `Ok::<_, Error>(())`
  - `Ok(assert_eq!(foo, bar)) as Result<(), Error>` ->
    `assert_eq!(foo, bar); Ok::<_, Error>(())`

- `management`

  - Warnings for unused code and dead code are no longer disabled.
  - `APIClient` no longer holds `Arc<Configuration<C>>` since it does not use it.
    Thus it also no longer needs to be generic on `C`
  - Rename `Error::ApiError` variant to `Error::Api`

- `provisioning`

  - Some `static` `&'static str` changed to `const`
  - `Ok(()) as Result<(), Error>` -> `Ok::<_, Error>(())`

- `tokio-named-pipe`

  - Avoid potential overflow converting u64 to u32.

- `workload`

  - Warnings for unused code and dead code are no longer disabled.
  - `APIClient` no longer holds `Arc<Configuration<C>>` since it does not use it.
    Thus it also no longer needs to be generic on `C`
  - Rename `Error::ApiError` variant to `Error::Api`

The following lints are disabled:

- `similar_names`: Over-zealous.
- `stutter`: Complains about a type named `ErrorKind` inside a module
  named `error`.
- `use_self`: Complains about using the type's name in struct literals,
  which is not a common practice.

### CODE REVIEW

File: edgelet/edgelet-docker/src/runtime.rs [hunk 1]
```diff
@@ -593,19 +602,17 @@ mod tests {
...
+        assert_eq!(0, DockerModuleRuntime::merge_env(cur_env, &new_env).len());
     }
 
     #[test]
     fn merge_env_new_empty() {
         let cur_env = Some(vec!["k1=v1".to_string(), "k2=v2".to_string()]);
         let new_env = HashMap::new();
-        let mut merged_env = DockerModuleRuntime::merge_env(cur_env.as_ref(), &new_env);
+        let mut merged_env =
+            DockerModuleRuntime::merge_env(cur_env.as_ref().map(AsRef::as_ref), &new_env);
```
Comment (damonbarry) [inline edgelet/edgelet-docker/src/runtime.rs]: Is it necessary to call `as_ref()` _AND_ `map(AsRef::as_ref)` on `cur_env` here?

Comment (arsing) [reply edgelet/edgelet-docker/src/runtime.rs]: `cur_env` is `Option<Vec<String>>` and `merge_env` wants `Option<&[String]>`. `Option::as_ref` is `&Option<T> -> &Option<T>`. `AsRef::as_ref` is `&T -> &U`

File: edgelet/edgelet-docker/src/module.rs [hunk 2]
```diff
@@ -69,40 +65,38 @@ impl<C: 'static + Connect> Module for DockerModule<C> {
                 .container_inspect(&self.name, false)
                 .map(|resp| {
                     resp.state()
-                        .map(|state| {
+                        .map_or_else(ModuleRuntimeState::default, |state| {
```
[the remark this answers is not in the document]
Comment (arsing) [reply edgelet/edgelet-docker/src/module.rs]: This is `Option`, for which `map_or_else` has existed for a long time. It was somewhat redundant for `Result` which is why that was only added recently (and not yet stabilized).

File: edgelet/edgelet-docker/src/runtime.rs (hunk 1 above)
Comment (damonbarry) [reply edgelet/edgelet-docker/src/runtime.rs]: Got it, thanks

File: edgelet/edgelet-test-utils/src/identity.rs [hunk 3]
```diff
@@ -190,8 +190,9 @@ impl IdentityManager for TestIdentityManager {
         self.identities
             .iter()
             .position(|ref mid| mid.module_id() == id.module_id())
-            .map(|index| self.identities.remove(index))
-            .map(|_| future::ok(()))
-            .unwrap_or_else(|| future::err(Error::ModuleNotFound))
+            .map(|index| {
+                self.identities.remove(index);
+            }).ok_or(Error::ModuleNotFound)
```
[the remark this answers is not in the document]
Comment (arsing) [reply edgelet/edgelet-test-utils/src/identity.rs]: The result of the `map` is an `Option<()>`, yes. The old code did this via two `map`s where the second one mapped to a `Future<()>`. The new code builds a `Result<()>` and converts it to a `Future` with `IntoFuture::into_future`

File: edgelet/management/src/apis/client.rs [hunk 4]
```diff
@@ -3,19 +3,20 @@ use std::sync::Arc;
...
+pub struct APIClient {
     identity_api: Box<::apis::IdentityApi>,
     module_api: Box<::apis::ModuleApi>,
     system_information_api: Box<::apis::SystemInformationApi>,
 }
 
-impl<C: hyper::client::connect::Connect + 'static> APIClient<C> {
-    pub fn new(configuration: Configuration<C>) -> APIClient<C> {
+impl APIClient {
+    pub fn new<C>(configuration: Configuration<C>) -> Self
```
Comment (damonbarry) [inline edgelet/management/src/apis/client.rs]: Nice!

**Comment by arsing** [comment]
@damonbarry @darobs @mrohera The merge primarily effected [edgelet/edgelet-http-workload/src/server/cert/identity.rs](https://github.com/Azure/iotedge/pull/515/commits/70d5974e569b56eb77c0dc46e2b29d329f26a7e0#diff-74d26d6cb427b63a2f4eb74d246d378c) and [ edgelet/edgelet-http-workload/src/server/cert/server.rs](https://github.com/Azure/iotedge/pull/515/commits/70d5974e569b56eb77c0dc46e2b29d329f26a7e0#diff-aeccb2659936208dd45d5b09373a96e0)

Can you review and approve again?

**Comment by mrohera** [comment]
@arsing I reviewed the files called out and LGTM.

**Pull request merged**